### PR TITLE
new module DeepMAsED

### PR DESCRIPTION
**Pull request by SkyLexS**
## PR checklist

- [x] DeepMAsED features extracts alignment-based statistics from a BAM file and assembly FASTA to produce per-contig feature tables, which are then consumed by DeepMAsED predict to assign each contig a misassembly score between 0 (correctly assembled) and 1 (likely misassembly) using a pre-trained neural network.
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [x] If necessary, include test data in your PR.
- [x] Remove all TODO statements.
- [x] Broadcast software version numbers to `topic: versions` - [See version_topics](https://nf-co.re/blog/2025/version_topics)
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - For modules:
    - [x] `nf-core modules test <MODULE> --profile docker`
    - [ ] `nf-core modules test <MODULE> --profile singularity`
    - [ ] `nf-core modules test <MODULE> --profile conda`
  - For subworkflows:
    - [x] `nf-core subworkflows test <SUBWORKFLOW> --profile docker`
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile singularity`
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile conda`

### CODE REVIEW

File: modules/nf-core/deepmased/features/tests/main.nf.test [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true),
+                ]
+                """
+            }
+        }
+
+        then {
+            assertAll(
+                { assert process.success },
+                { assert snapshot(process.out).match() }
```
Comment (famosab) [inline modules/nf-core/deepmased/features/tests/main.nf.test]:
```suggestion
                { assert snapshot(sanitizeOutput(process.out)).match() }
```

You can use that for a nicer snapshot

File: modules/nf-core/deepmased/features/environment.yml [hunk 2]
```diff
@@ -0,0 +1,9 @@
+---
+# yaml-language-server: $schema=https://raw.githubusercontent.com/nf-core/modules/master/modules/environment-schema.json
+channels:
+  - conda-forge
+  - bioconda
+dependencies:
+  # renovate: datasource=conda depName=bioconda/deepmased
+  - bioconda::deepmased=0.3.1
+  - conda-forge::setuptools=78.1
```
Comment (famosab) [inline modules/nf-core/deepmased/features/environment.yml:9]: if we pin it here I would also add this to the versions export

File: modules/nf-core/deepmased/features/main.nf [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    conda "${moduleDir}/environment.yml"
+    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
+        'https://depot.galaxyproject.org/singularity/deepmased:0.3.1--pyh5ca1d4c_0':
+        'quay.io/biocontainers/deepmased:0.3.1--pyh5ca1d4c_0' }"
+
+    input:
+    tuple val(meta), path(bam), path(bai), path(fasta)
+
+    output:
+    tuple val(meta), path("${prefix}_feature_file_paths.tsv"), path("*_feats.tsv"), emit: features
```
Comment (famosab) [inline modules/nf-core/deepmased/features/main.nf]: Please emit one file per output tuple!

File: modules/nf-core/deepmased/features/main.nf [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    prefix     = task.ext.prefix ?: "${meta.id}"
+    """
+    echo -e "bam\\tfasta" > ${prefix}_file_paths.tsv
+    echo -e "${bam}\\t${fasta}" >> ${prefix}_file_paths.tsv
+
+    DeepMAsED features \\
+        ${prefix}_file_paths.tsv \\
+        -p ${task.cpus} \\
+        -o . \\
+        -n ${prefix}_feature_file_paths.tsv \\
```
Comment (famosab) [inline modules/nf-core/deepmased/features/main.nf:37]: If input and output names can overlap, use `task.ext.prefix` to disambiguate. For example:
```
if ("${tsv}" == "${prefix}.tsv") error "Input and output names are the same, use \"task.ext.prefix\" to disambiguate!"
```

File: modules/nf-core/deepmased/features/meta.yml [hunk 5]
```diff
@@ -0,0 +1,88 @@
+name: "deepmased_features"
+description: "DeepMAsED features subcommand: extracts alignment-based features from
```
Comment (famosab) [inline modules/nf-core/deepmased/features/meta.yml:2]:
```suggestion
description: "Extracts alignment-based features from
```

File: modules/nf-core/deepmased/features/meta.yml [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+        type: file
+        description: Sorted BAM file of reads mapped to the assembly
+        pattern: "*.{bam}"
+        ontologies:
+          - edam: "http://edamontology.org/format_2572"
+    - bai:
+        type: file
+        description: BAM index file
+        pattern: "*.{bai}"
+        ontologies: []
```
Comment (famosab) [inline modules/nf-core/deepmased/features/meta.yml]: ontology?

File: modules/nf-core/deepmased/predict/tests/main.nf.test [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+                    file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.paired_end.sorted.bam', checkIfExists: true),     // dummy feature_files
+                ]
+                """
+            }
+        }
+
+        then {
+            assertAll(
+                { assert process.success },
+                { assert snapshot(process.out).match() }
```
Comment (famosab) [inline modules/nf-core/deepmased/predict/tests/main.nf.test]:
```suggestion
                { assert snapshot(sanitizeOutput(process.out)).match() }
```

File: modules/nf-core/deepmased/predict/environment.yml [hunk 8]
```diff
@@ -0,0 +1,9 @@
+---
+# yaml-language-server: $schema=https://raw.githubusercontent.com/nf-core/modules/master/modules/environment-schema.json
+channels:
+  - conda-forge
+  - bioconda
+dependencies:
+  # renovate: datasource=conda depName=bioconda/deepmased
+  - bioconda::deepmased=0.3.1
+  - conda-forge::setuptools=78.1
```
Comment (famosab) [inline modules/nf-core/deepmased/predict/environment.yml:9]: see comment above

File: modules/nf-core/deepmased/predict/main.nf [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    task.ext.when == null || task.ext.when
+
+    script:
+    def args   = task.ext.args ?: '--cpu-only'
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    """
+    DeepMAsED predict \\
+        ${feature_file_table} \\
+        --n-procs ${task.cpus} \\
+        --save-name ${prefix}_deepmased \\
```
Comment (famosab) [inline modules/nf-core/deepmased/predict/main.nf]: We want to let the module user decide about the output file naming in their pipeline via `ext-prefix`. So ideally the output filenames are just `${prefix}`.

File: modules/nf-core/deepmased/predict/meta.yml [hunk 10]
```diff
@@ -0,0 +1,79 @@
+name: "deepmased_predict"
+description: "DeepMAsED predict subcommand: runs the pre-trained deep learning model
```
Comment (famosab) [inline modules/nf-core/deepmased/predict/meta.yml:2]:
```suggestion
description: "Runs the pre-trained deep learning model
```

**Review by famosab — COMMENTED** [review]
Looks godd :) I added a few comments to finalize this!

**Comment by SkyLexS** [comment]
About the comments you've added.

I chose to keep versions.yml as a single file instead of topic: versions because it works much better with the mag pipeline and is easier to maintain (I am currently testing deepmased within mag to be sure it works properly before committing here. The snapshots in the tests are auto-generated by nf-test and manual reformatting would break their JSON serialization, so I left the format that the tool produces. And for the descriptions in meta.yml, I preferred to leave the more detailed version that is synchronized with the mag implementation, because it helps users better understand what these modules do.

if you have any tips and/or tricks, please tell

**Comment by famosab** [comment]
> I chose to keep versions.yml as a single file instead of topic: versions because it works much better with the mag pipeline and is easier to maintain (I am currently testing deepmased within mag to be sure it works properly before committing here. 

We are swapping all modules to topics, the versions.yml will be deprectated in the future. Even if mag still uses it now, as soon as the new template update is merged topics will be as easy to use as the versions.yml so please swap to the newer syntax!

> The snapshots in the tests are auto-generated by nf-test and manual reformatting would break their JSON serialization, so I left the format that the tool produces. 

I am not sure what you are referring to. I added no comments on the snapshots just some hints how to make them more human readable by editing the main.nf.test files and then regenerating the snapshots, which is what I would still suggest to do.

> And for the descriptions in meta.yml, I preferred to leave the more detailed version that is synchronized with the mag implementation, because it helps users better understand what these modules do.

Feel free to leave it how you want it. I just thought its redundant information because one line above we see `deepmased_predict` so its clear that this module is focused on the `predict` subcommand. But thats just my opinion and not a blocker!

File: modules/nf-core/deepmased/features/main.nf [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+    tuple val("${task.process}"), val('deepmased'), val('0.3.1'), emit: versions_deepmased, topic: versions
+    tuple val("${task.process}"), val('setuptools'), val('78.1') , emit: versions_setuptools, topic: versions
+
+    when:
+    task.ext.when == null || task.ext.when
+
+    script:
+    def args   = task.ext.args ?: ''
+    prefix     = task.ext.prefix ?: "${meta.id}"
+    def VERSION = '0.3.1' // WARN: Version information not provided by tool on CLI. Please update this string when bumping container versions.
```
Comment (jfy133) [inline modules/nf-core/deepmased/features/main.nf]: Can be removed as hardocded in `output` and you have the warning above the containers already
```suggestion
```

File: modules/nf-core/deepmased/predict/main.nf [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+    tuple val("${task.process}"), val('deepmased'), val('0.3.1'), emit: versions_deepmased, topic: versions
+    tuple val("${task.process}"), val('setuptools'), val('78.1') , emit: versions_setuptools, topic: versions
+
+    when:
+    task.ext.when == null || task.ext.when
+
+    script:
+    def args   = task.ext.args ?: ''
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    def VERSION = '0.3.1' // WARN: Version information not provided by tool on CLI. Please update this string when bumping container versions.
```
Comment (jfy133) [inline modules/nf-core/deepmased/predict/main.nf]: Same as above
```suggestion
```

File: modules/nf-core/deepmased/features/main.nf [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+    """
+    # Check for input/output name collision
+    if [[ "${prefix}_file_paths.tsv" == "${prefix}_feature_file_paths.tsv" ]]; then
+        echo "ERROR: Input TSV filename matches output filename. Set ext.prefix differently." >&2
+        exit 1
+    fi
```
Comment (jfy133) [inline modules/nf-core/deepmased/features/main.nf]: This should check should go before the bash itself as we want the error to happe nat nextflow not bash script execution: https://nf-co.re/docs/specifications/components/modules/naming-conventions#command-file-output-naming

File: modules/nf-core/deepmased/features/tests/main.nf.test [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+            }
+        }
+
+        then {
+            assertAll(
+                { assert process.success },
+                { assert snapshot(
+                    process.out.feature_table,
+                    process.out.feature_files,
+                    process.out.findAll { key, val -> key.startsWith("versions") }
```
Comment (jfy133) [inline modules/nf-core/deepmased/features/tests/main.nf.test]: You should switch to to the 'sanitizeOutput' thing that Famke referred to above, a lot of modules are using the mas an exmaple.

**Review by jfy133 — COMMENTED** [review]
As a general point from reading the discussion: for the future - modules should not be adapted for specific pipelines, and should be independent and usable fr all. So don't modify a module away from the module specifications to 'satisfy' the pipeline ;).

Almost there, ocne comments below addressed, I thin kyou will have satisfied Famke's comments :)

**Review by jfy133 — APPROVED** [review]

File: modules/nf-core/deepmased/features/tests/main.nf.test [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+                """
+            }
+        }
+
+        then {
+            assertAll(
+                { assert process.success },
+                { assert snapshot(
+                    process.out.feature_table,
+                    process.out.feature_files,
```
Comment (famosab) [inline modules/nf-core/deepmased/features/tests/main.nf.test]:
```suggestion
```

This is not needed because it just duplicates the info then

File: modules/nf-core/deepmased/features/tests/main.nf.test [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
```
Comment (famosab) [inline modules/nf-core/deepmased/features/tests/main.nf.test]:
```suggestion
```
same here

File: modules/nf-core/deepmased/predict/tests/main.nf.test [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+                ]
+                """
+            }
+        }
+
+        then {
+            assertAll(
+                { assert process.success },
+                { assert snapshot(
+                    process.out.predictions,
```
Comment (famosab) [inline modules/nf-core/deepmased/predict/tests/main.nf.test]:
```suggestion
```

File: modules/nf-core/deepmased/predict/main.nf [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+    script:
+    def args   = task.ext.args ?: ''
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    """
+    DeepMAsED predict \\
+        ${feature_file_table} \\
+        --n-procs ${task.cpus} \\
+        --save-name ${prefix} \\
+        ${args}
+
```
Comment (famosab) [inline modules/nf-core/deepmased/predict/main.nf]:
```suggestion
```

**Review by famosab — APPROVED** [review]
